doc/compiler/il/Node.md

```diff
@@ -0,0 +1,91 @@
+
+Normally we would create a Node like this:
+
+```cpp
+TR::Node *node = TR::Node::create(...);
+```
+
+and `create` would create a new Node from the NodePool like this:
+
+```cpp
+TR::Node *node = new (comp->getNodePool()) TR::Node(...);
+```
+
+where the node pool allocator would add a new Node entry to the NodePool table, and return a reference to that entry.
+
+There also exists an alternative Node interface; a new node can be created over the original Node using `recreateWithoutProperties`:
+
+```cpp
+TR::Node *recreatedNode = TR::Node::recreateWithoutProperties(originalNode, ...);
+```
+
+The returned value (the address of the recreated node) should be identical to the address of the original node passed to recreate. The old node can be recreated like this (we may also need to "destroy" originalNode without deallocating its memory from the node pool):
+
+```cpp
+originalIndex = originalNode->getNodePoolIndex();
+TR::Node *node = new (comp->getNodePool(), originalIndex) TR::Node(...);
+```
+
+Here, we provide an alternate form of the `new` operator. It has an extra parameter: the index of the original Node that we wish to allocate on top of. In this case, instead of adding an entry to the table, the allocator just checks the index and returns the original storage, which is cleared and initialized normally.
+
+When the immutable recreate interface is used, the properties of the new Node need to be explicitly set. Anything not explicitly set in implicitly unset. This will lead to code that is more explicit about what it means to do.
+
+### Recreate functions that do not copy properties
+
+When the replacement node bears no or little relationship to the original node, little or no properties should be copied to the new node. Thus some recreate functions do not copy any properties, children or symRefs, but they do recreate the node with new children and new symRefs. These are:
+
+```cpp
+static TR::Node *recreateWithoutProperties(TR::Node *originalNode, OMR::ILOpCodes op, uint16_t numChildren, [TR::Node *first, ... ]);
+static TR::Node *recreateWithSymRefWithoutProperties(TR::Node *originalNode, OMR::ILOpCodes op, uint16_t numChildren, uint16_t numChildArgs, TR::Node *first, ChildrenAndSymRefType... childrenAndSymRef);
+```
+
+### Recreate functions that do copy properties
+
+In practice, the above type of mutation is rare. In most cases the replacement node has a close relationship to the original node. Thus the more common interface is this:
+
+```cpp
+static TR::Node *recreate(TR::Node *originalNode, OMR::ILOpCodes op);
+```
+
+The function `recreate` recreates a replacement node with the new opcode, but with all the original properties, children and symRefs as the original as long as they may be valid for the new opcode. This was a common way of using the deprecated function `setOpCodeValue` in TR, but unlike `setOpCodeValue`, `recreate` will clear invalid properties, which avoids some of the issues that happened with `setOpCodeValue`.
```

Comment:
* Add a comment after 'children'

* I'd change the last sentence to read something like:

Unlike `setOpCodeValue`, `recreate` will clear invalid properties.